src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -2547,67 +2541,61 @@ internal static int IndexOfAnyValueType<T>(ref T searchSpace, T value0, T value1
                 }
-                while (!Unsafe.IsAddressLessThan(ref currentSearchSpace, ref searchSpace));
 
-                // If any elements remain, process the first vector in the search space.
-                if ((uint)length % Vector256<TValue>.Count != 0)
+                // Process the first vector in the search space.
+
+                current = Vector256.LoadUnsafe(ref searchSpace);
+                equals = TNegator.NegateIfNeeded(Vector256.Equals(current, values0) | Vector256.Equals(current, values1) | Vector256.Equals(current, values2) | Vector256.Equals(current, values3));
+
+                if (equals != Vector256<TValue>.Zero)
                 {
-                    current = Vector256.LoadUnsafe(ref searchSpace);
-                    equals = TNegator.NegateIfNeeded(Vector256.Equals(current, values0) | Vector256.Equals(current, values1)
-                                            | Vector256.Equals(current, values2) | Vector256.Equals(current, values3));
-                    if (equals != Vector256<TValue>.Zero)
-                    {
-                        return ComputeLastIndex(ref searchSpace, ref searchSpace, equals);
-                    }
+                    return ComputeLastIndex(offset: 0, equals);
                 }
             }
             else
             {
                 Vector128<TValue> equals, current, values0 = Vector128.Create(value0), values1 = Vector128.Create(value1), values2 = Vector128.Create(value2), values3 = Vector128.Create(value3);
-                ref TValue currentSearchSpace = ref Unsafe.Add(ref searchSpace, length - Vector128<TValue>.Count);
+                nint offset = length - Vector128<TValue>.Count;
 
                 // Loop until either we've finished all elements or there's less than a vector's-worth remaining.
-                do
+                while (offset > 0)
                 {
-                    current = Vector128.LoadUnsafe(ref currentSearchSpace);
-                    equals = TNegator.NegateIfNeeded(Vector128.Equals(current, values0) | Vector128.Equals(current, values1)
-                                            | Vector128.Equals(current, values2) | Vector128.Equals(current, values3));
+                    current = Vector128.LoadUnsafe(ref searchSpace, (nuint)(offset));
+                    equals = TNegator.NegateIfNeeded(Vector128.Equals(current, values0) | Vector128.Equals(current, values1) | Vector128.Equals(current, values2) | Vector128.Equals(current, values3));
+
                     if (equals == Vector128<TValue>.Zero)
                     {
-                        currentSearchSpace = ref Unsafe.Subtract(ref currentSearchSpace, Vector128<TValue>.Count);
+                        offset -= Vector128<TValue>.Count;
                         continue;
                     }
 
-                    return ComputeLastIndex(ref searchSpace, ref currentSearchSpace, equals);
+                    return ComputeLastIndex(offset, equals);
```

Comment:
It doesn't affect this PR, but I wonder why it's written like this rather than:
```C#
if (equals != Vector128<TValue>.Zero)
{
    return ComputeLastIndex(offset, equals);
}
```
Is this just style, or we expect the not found case to be the most common and want the branch predictor to default to that?

Reply:
> we expect the not found case to be the most common and want the branch predictor to default to that

That was my assumption when I was implementing it.